### ISSUE TITLE
Consider not computing the revert message if revert condition is not met

### ISSUE THREAD

**Issue by Amxx** [ethereum/solidity#11201]
## Abstract

I noticed that when adding a `require` statement with a complex revert reason (computed by an internal function), I have to pay the cost of computing this revert string regardless of the value of the precondition. Also, if that internal function is not view, and include some state changes, these changes are enacted event when the condition does require a revert.

Here is an example of code
```
contract test {
    uint256 public i = 0;
    
    function trigger() external {
        require(true, getErrorMessage());
    }
    
    function getErrorMessage() internal returns (string memory) {
        i++;
        return "";
    }
}
```

When calling trigger, I was not expecting the `getErrorMessage()` to be called, and I am not expecting `i` to be incremented. In a way, I'm expecting:
`requier(condition_or_call, string_or_call);` to be translated into `if (!condition_or_call) revert(string_or_call);`

[In this particular case, using `if-revert` saves about 30k gas on valid paths that do not revert](https://github.com/Amxx/openzeppelin-contracts/blob/feature/AccessControlRevertMessage/contracts/access/AccessControl.sol#L122-L131)

## Motivation

Having the compiler rewrite `require()` using this `if() revert()` logic would reduce the gas cost of non reverting paths since the complex error message would not have to be computer unless absolutely necessary. Also, it means the revert string can be computed with additional assumptions.

## Specification

Internally (yul?) rewrite all `require(x,y)` statements as `if (!x) revert(y);`

## Backwards Compatibility

Will this does change the behavior of the produced bytecode, I believe it does in a way that is compliant with the common understanding of `require(bool,string)` by devs. Still, this might be breaking in some extreme cases that have to be studied.

**Comment by axic** [comment]
This is a known "issue". The `revert` function is treated as a regular function and as such its arguments are evaluated prior to invoking it.

In #7877 we are discussing potential future improvements to this. The consensus so far was that changing the behaviour of `revert` alone would be weird, but we plan to introduce new functionality which makes this more clear and obsoletes the old one.

**Comment by cameel** [comment]
I think that the expectation comes from the fact that until very recently (#10089) our docs stated that they are equivalent. Now we have a caveat about this in [the docs for `revert()`](https://docs.soliditylang.org/en/latest/control-structures.html#revert):

> The `require` function is evaluated just as any other function. This means that all arguments are evaluated before the function itself is executed. In particular, in `require(condition, f())` the function `f` is executed even if `condition` is true.

I don't think that just changing the semantics of `require()` without changing the syntax is a good idea. It looks like a function call and to me  for example it would be completely unintuitive if side-effects did not to happen when the condition is false. And on top of that it's been working like that for a long time so it has a big potential to silently break code that already assumes the correct behavior.

But I agree that conflicting expectations here are a problem and it's actually one of the reasons (though not the only one) why we haven't extended `require()` to support the new custom error system that will be released in 0.8.4 (see the issue @axic linked to; [this comment](https://github.com/ethereum/solidity/issues/7877#issuecomment-780598691) is a good summary of the current situation). For now only the newly introduced `revert` statement will support custom errors. For `require()` we'd prefer to change the syntax to make it clearer but all the proposals so far have some downsides and it's possible that we will end up just deprecating `require()` and recommending the `if (!c) revert Error();` pattern instead.

**Comment by axic** [comment]
> recommending the if (!c) revert Error(); pattern instead.

Or `unless (c) revert Error();` :wink:

But I think it would be early to make such statements. `require` is also useful for other subsystems, like the SMTChecker, that does not mean it won't be able to understand a if/revert unless/revert statement, but it is something we need to take into consideration.

**Comment by cameel** [comment]
I'd count that as new syntax :)

Just to be clear, I'm not saying we will definitely deprecate it. Just that we could not agree on a replacement yet and even the new docs use the `if (!c) revert Error();` pattern for it so it's a possibility. Sorry if I worded it in a way that implies that it's more likely than it actually is.

**Comment by sakulstra** [comment]
I know this was closed a long time ago, but i initially missed it and opened the discussion here: https://forum.soliditylang.org/t/eager-require-evaluation-is-unexpected/3744

I did a bit of "research" (checking popular opensource contracts in the dex and lending space) and my observation is that the current behavior is always by accident. While i did not find a single occasion of intended side effects in the `string_or_call` argument, I found numerous usages of external calls, storage access and memory access that **quite clearly are unintended**'

I did a very AI driven(given I have zero experience with the compiler) exploration of a `if (!c) revert Error();` approach, just to see how it affects projects.
As kinda expected for the projects i benchmarked:
- gas reduces measurably
- for projects with a lot of errors, bytecode increases measurably <- i think this is as now the code is optimized differently

---

**quite clearly are unintended**': while in most cases i was not able to ask the authors, I think it's save to assume that ppl who use bitmaps, and assembly to optimize code in some places, do not willingly throw away gas on a eager evaluation on a non-met error condition.

---

Personally I think this should be reconsidered.
